File: lua/levels.lua
local json = require "libs.json"
local DataFixer = require "datafixer"
local dbg = require "util.dbg"

local Levels = {}

local fileName = "levels.json"
local levels

local function init()
    if love.filesystem.getInfo(fileName) then
        levels = json.decode(love.filesystem.read(fileName))
        dbg.printShallowSequence(levels)
        if #levels > 0 then 
            for i = #levels, 1, -1 do
                if DataFixer.fixData(levels[i]) == false then
                    table.remove(levels, i)
                end
            end
        end
    else
        levels = {}
    end
end

local function save()
    love.filesystem.write(fileName, json.encode(levels))
end

Levels.addLevel = function(level)
    table.insert(levels, level)
    save()
end

Levels.removeLevel = function(levelIndex)
    table.remove(levels, levelIndex)
    save()
end

Levels.nameUsed = function(name)
    for i = 1, #levels do
        if levels[i].name == name then
            return true
        end
    end
    return false
end

Levels.getLevels = function()
    return levels
end

init()
save()

return Levels

File: src/levels.ts
import { Level } from "./levelbuilder";
import { fixData } from "./datafixer";

interface RxiJsonLibrary {
    decode(this:void, input: string): any;
    encode(this:void, obj: any): string;
}

const json: RxiJsonLibrary = require("libs.json");

const fileName = "levels.json";
let levels: Level[];

function init() {
    if (love.filesystem.getInfo(fileName)) {
        const [contents, size] = love.filesystem.read(fileName);
        if (contents) {
            levels = json.decode(contents);
            levels = levels.filter((level) => fixData(level));
        } else {
            levels = [];
        }
    } else {
        levels = [];
    }
}

function save() {
    love.filesystem.write(fileName, json.encode(levels) as any);
}

init();
save();

export namespace Levels {
    export function addLevel(level: Level) {
        levels.push(level);
        save();
    }

    export function removeLevel(index: number) {
        table.remove(levels, index);
        save();
    }

    export function nameUsed(name: string): boolean {
        for (const level of levels) {
            if (level.name == name) {
                return true;
            }
        }
        return false;
    }

    export function getLevels(): Level[] {
        return levels;
    }
}